### PR TITLE
Address various DEM testing issues

### LINKED ISSUE #34297: [CERT-TEST-FAILURE] - TC-DEM-1.1 - ClusterRevision value has to be updated for Device Energy Management Cluster
### Feature Area

Other

### Test Case

TC-DEM-1.1

### Reproduction steps

1. Bring up the DUT with energy-management-app.
2. Pair the DUT using chip-tool through "Onnetwork" transport.
3. Run the command `./chip-tool deviceenergymanagement read cluster-revision 1 1`

**Expected Outcome:**

> [1720690440.539] [50435:50437] [TOO] Endpoint: 1 Cluster: 0x0000_0098 Attribute 0x0000_FFFD DataVersion: 521028924
> [1720690440.539] [50435:50437] [TOO]   ClusterRevision: 4
> [1720690440.540] [50435:50437] [EM] <<< [E:5099i S:17862 M:203971832 (Ack:69830139)] (S) Msg TX to 1:0000000000000001 [C5DD] [UDP:[fe80::e65f:1ff:fe0f:2878%eth0]:5540] --- Type 0000:10 (SecureChannel:StandaloneAck)

**Actual Outcome:**

> [1720690440.539] [50435:50437] [TOO] Endpoint: 1 Cluster: 0x0000_0098 Attribute 0x0000_FFFD DataVersion: 521028924
> [1720690440.539] [50435:50437] [TOO]   ClusterRevision: 3
> [1720690440.540] [50435:50437] [EM] <<< [E:5099i S:17862 M:203971832 (Ack:69830139)] (S) Msg TX to 1:0000000000000001 [C5DD] [UDP:[fe80::e65f:1ff:fe0f:2878%eth0]:5540] --- Type 0000:10 (SecureChannel:StandaloneAck)

### Bug prevalence

Whenever I try to read ClusterRevision.

### GitHub hash of the SDK that was being used

6e31453ed626768f9c74bc9e59ffae7a7137999f

### Platform

raspi

### Anything else?

In SPEC, the ClusterRevision value has been updated to 4. As per spec changes, SDK(energy-management-app) has to update the ClusterRevision value from 3 to 4.

**Specification References:** https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/DeviceEnergyManagement.adoc#revision-history

### LINKED ISSUE #34954: [Test Failed] DEM_TC_2_6 Device crashes.
### Test issue(s)

Device crashes at TH step for Test DEM_TC_2_6
`Executing Test Step: TH sends ModifyForecastRequest with ForecastID=Forecast.ForecastID+1, SlotAdjustments[0].{SlotIndex=0, Duration=Forecast.Slots[0].MaxDurationAdjustment}, Cause=GridOptimization.`

### Platform

_No response_

### Anything else?

[DEM2-6_2024_08_13_13_45_15.log](https://github.com/user-attachments/files/16600571/DEM2-6_2024_08_13_13_45_15.log)
[device.txt](https://github.com/user-attachments/files/16600572/device.txt)

### LINKED ISSUE #34960: [Test Failed] [TC_DEM_2_7] fails on ESP32
### Test issue(s)

Having applied a patch to resolve the crash seen on ESP32 - see this issue  https://github.com/project-chip/connectedhomeip/issues/34954

When running the TC_DEM_2_7 against the chip-energy-management-app we see that step 5 fails with Invalid Command, but the test script is expecting Constraint Error.

Further investigation is needed to understand why it is returning the wrong error code.

### Platform

esp32

### Anything else?

```

INFO       | 2024-08-13 18:15:54.353572 | Executing Test Step: TH reads Forecast attribute.
INFO       | 2024-08-13 18:15:54.625141 | Test Step Completed [PASSED]: TH reads Forecast attribute.
INFO       | 2024-08-13 18:15:54.627048 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:54.633160 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:54.634799 | Executing Test Step: TH reads OptOutState attribute.
INFO       | 2024-08-13 18:15:54.749669 | Test Step Completed [PASSED]: TH reads OptOutState attribute.
INFO       | 2024-08-13 18:15:54.751330 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:54.757734 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:54.759412 | Executing Test Step: TH sends RequestConstraintBasedPowerForecast with constraints[0].{StartTime=now()-10, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, Cause=LocalOptimization.
INFO       | 2024-08-13 18:15:55.174483 | Test Step Completed [PASSED]: TH sends RequestConstraintBasedPowerForecast with constraints[0].{StartTime=now()-10, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, Cause=LocalOptimization.
INFO       | 2024-08-13 18:15:55.176136 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:55.182476 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:55.184178 | Executing Test Step: TH sends RequestConstraintBasedPowerForecast with constraints[0].{StartTime=now()+10, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, constraints[1].{StartTime=now()+20, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, constraints[2].{StartTime=now()+40, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, Cause=LocalOptimization.
WARNING    | 2024-08-13 18:15:55.668233 | Test Failure: Python test step failure
INFO       | 2024-08-13 18:15:55.671687 | Test Step Completed [FAILED]: TH sends RequestConstraintBasedPowerForecast with constraints[0].{StartTime=now()+10, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, constraints[1].{StartTime=now()+20, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, constraints[2].{StartTime=now()+40, Duration=20, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, Cause=LocalOptimization.
INFO       | 2024-08-13 18:15:55.674015 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:55.683853 | --------------------------------------------------------------------------------
INFO       | 2024-08-13 18:15:55.685120 | Executing Test Step: Show test logs
INFO       | 2024-08-13 18:15:55.703813 | ---- Start of Python test logs ----
PYTHON_TEST | 2024-08-13 18:15:55.707039 | [MatterTest] 08-13 18:15:51.300 INFO ==========> TC_DEM_2_7 <==========
...

[MatterTest] 08-13 18:15:55.377 INFO <<< [E:42497i S:56886 M:131054175 (Ack:153280628)] (S) Msg TX to 1:0000000012344321 [B1F2] [UDP:172.24.0.102:5540] --- Type 0001:08 (IM:InvokeCommandRequest) (B:126)
[MatterTest] 08-13 18:15:55.541 INFO {
[MatterTest] 08-13 18:15:55.542 INFO 	"event" : "MessageReceived",
[MatterTest] 08-13 18:15:55.542 INFO 	"messageType" : "Secure",
[MatterTest] 08-13 18:15:55.543 INFO 	"packetHeader" : 
[MatterTest] 08-13 18:15:55.544 INFO 	{
[MatterTest] 08-13 18:15:55.544 INFO 		"flags" : 0,
[MatterTest] 08-13 18:15:55.545 INFO 		"msgCounter" : 153280629,
[MatterTest] 08-13 18:15:55.545 INFO 		"securityFlags" : 0,
[MatterTest] 08-13 18:15:55.546 INFO 		"sessionId" : 56886
[MatterTest] 08-13 18:15:55.547 INFO 	},
[MatterTest] 08-13 18:15:55.547 INFO 	"payload" : 
[MatterTest] 08-13 18:15:55.548 INFO 	{
[MatterTest] 08-13 18:15:55.548 INFO 		"decoded" : 
[MatterTest] 08-13 18:15:55.549 INFO 		{
[MatterTest] 08-13 18:15:55.549 INFO 			"invoke_response" : 
[MatterTest] 08-13 18:15:55.550 INFO 			{
[MatterTest] 08-13 18:15:55.550 INFO 				"interaction_model_revison" : "11",
[MatterTest] 08-13 18:15:55.551 INFO 				"invoke_responses" : 
[MatterTest] 08-13 18:15:55.551 INFO 				{
[MatterTest] 08-13 18:15:55.552 INFO 					"Anonymous<0>" : 
[MatterTest] 08-13 18:15:55.552 INFO 					{
[MatterTest] 08-13 18:15:55.553 INFO 						"status" : 
[MatterTest] 08-13 18:15:55.554 INFO 						{
[MatterTest] 08-13 18:15:55.554 INFO 							"path" : 
[MatterTest] 08-13 18:15:55.555 INFO 							{
[MatterTest] 08-13 18:15:55.555 INFO 								"cluster_id" : "152 == 'DeviceEnergyManagement'",
[MatterTest] 08-13 18:15:55.556 INFO 								"command_id" : "6 == 'RequestConstraintBasedForecast'",
[MatterTest] 08-13 18:15:55.556 INFO 								"endpoint_id" : "1"
[MatterTest] 08-13 18:15:55.557 INFO 							},
[MatterTest] 08-13 18:15:55.557 INFO 							"status" : 
[MatterTest] 08-13 18:15:55.558 INFO 							{
[MatterTest] 08-13 18:15:55.558 INFO 								"status" : "133 == kInvalidCommand"
[MatterTest] 08-13 18:15:55.559 INFO 							}
[MatterTest] 08-13 18:15:55.559 INFO 						}
[MatterTest] 08-13 18:15:55.560 INFO 					}
[MatterTest] 08-13 18:15:55.560 INFO 				},
[MatterTest] 08-13 18:15:55.561 INFO 				"suppress_response" : "false"
[MatterTest] 08-13 18:15:55.561 INFO 			}
[MatterTest] 08-13 18:15:55.562 INFO 		},
[MatterTest] 08-13 18:15:55.562 INFO 		"hex" : "152800360115350137002400012401982402061835012400851818181824FF0B18",
[MatterTest] 08-13 18:15:55.563 INFO 		"size" : 33
[MatterTest] 08-13 18:15:55.563 INFO 	},
[MatterTest] 08-13 18:15:55.564 INFO 	"payloadHeader" : 
[MatterTest] 08-13 18:15:55.564 INFO 	{
[MatterTest] 08-13 18:15:55.565 INFO 		"ackMessageCounter" : 131054175,
[MatterTest] 08-13 18:15:55.565 INFO 		"exchangeFlags" : 6,
[MatterTest] 08-13 18:15:55.566 INFO 		"exchangeId" : 42497,
[MatterTest] 08-13 18:15:55.566 INFO 		"initiator" : false,
[MatterTest] 08-13 18:15:55.566 INFO 		"messageType" : 9,
[MatterTest] 08-13 18:15:55.567 INFO 		"needsAck" : true,
[MatterTest] 08-13 18:15:55.567 INFO 		"protocolId" : 1
[MatterTest] 08-13 18:15:55.568 INFO 	}
[MatterTest] 08-13 18:15:55.568 INFO }
[MatterTest] 08-13 18:15:55.569 INFO >>> [E:42497i S:56886 M:153280629 (Ack:131054175)] (S) Msg RX from 1:0000000012344321 [B1F2] --- Type 0001:09 (IM:InvokeCommandResponse) (B:67)
[MatterTest] 08-13 18:15:55.582 INFO Received Command Response Status for Endpoint=1 Cluster=0x0000_0098 Command=0x0000_0006 Status=0x85
[MatterTest] 08-13 18:15:55.585 ERROR Exception occurred in test_TC_DEM_2_7.
Traceback (most recent call last):
  File "/root/python_testing/scripts/sdk/TC_DEMTestBase.py", line 162, in send_request_constraint_based_forecast
    await self.send_single_cmd(cmd=Clusters.DeviceEnergyManagement.Commands.RequestConstraintBasedForecast(constraints=constraintList,
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 992, in send_single_cmd
    result = await dev_ctrl.SendCommand(nodeid=node_id, endpoint=endpoint, payload=cmd, timedRequestTimeoutMs=timedRequestTimeoutMs,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/chip/ChipDeviceCtrl.py", line 1213, in SendCommand
    return await future
           ^^^^^^^^^^^^
chip.interaction_model.InteractionModelError: InteractionModelError: InvalidCommand (0x85)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/mobly/base_test.py", line 818, in exec_one_test
    test_method()
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 1671, in async_runner
    return _async_runner(body, self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 1659, in _async_runner
    return asyncio.run(runner_with_timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
    return await fut
           ^^^^^^^^^
  File "/root/python_testing/scripts/sdk/TC_DEM_2_7.py", line 194, in test_TC_DEM_2_7
    await self.send_request_constraint_based_forecast(constraintList, cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization, expected_status=Status.ConstraintError)
  File "/root/python_testing/scripts/sdk/TC_DEMTestBase.py", line 170, in send_request_constraint_based_forecast
    asserts.assert_equal(e.status, expected_status, "Unexpected error returned")
  File "/usr/local/lib/python3.12/dist-packages/mobly/asserts.py", line 66, in assert_equal
    _call_unittest_assertion(
  File "/usr/local/lib/python3.12/dist-packages/mobly/asserts.py", line 50, in _call_unittest_assertion
    raise signals.TestFailure(my_msg, extras=extras)
mobly.signals.TestFailure: Details=<Status.InvalidCommand: 133> != <Status.ConstraintError: 135> Unexpected error returned, Extras=None
```

### LINKED ISSUE #34962: [Test Failed] [TC_DEM_2.8] Step 5 - Invalidcommand returned instead on ConstraintError
### Test issue(s)

As part of TE2 Matter 1.4 testing on an embedded target, we ran with a patch to fix a crash see: (https://github.com/project-chip/connectedhomeip/issues/34954).

When running TC_DEM_2.8 we got a similar error as seen in TC_DEM_2.7 https://github.com/project-chip/connectedhomeip/issues/34960 on step 5.

### Platform

core (please add to version below)

### Anything else?

`[MatterTest] 08-13 18:40:42.009 INFO >>> [E:63143i S:16393 M:23250085 (Ack:201973772)] (S) Msg RX from 1:0000000012344321 [1CE3] --- Type 0001:09 (IM:InvokeCommandResponse) (B:67)
[MatterTest] 08-13 18:40:42.022 INFO Received Command Response Status for Endpoint=1 Cluster=0x0000_0098 Command=0x0000_0006 Status=0x85
[MatterTest] 08-13 18:40:42.025 ERROR Exception occurred in test_TC_DEM_2_8.
Traceback (most recent call last):
  File "/root/python_testing/scripts/sdk/TC_DEMTestBase.py", line 162, in send_request_constraint_based_forecast
    await self.send_single_cmd(cmd=Clusters.DeviceEnergyManagement.Commands.RequestConstraintBasedForecast(constraints=constraintList,
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 992, in send_single_cmd
    result = await dev_ctrl.SendCommand(nodeid=node_id, endpoint=endpoint, payload=cmd, timedRequestTimeoutMs=timedRequestTimeoutMs,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/chip/ChipDeviceCtrl.py", line 1213, in SendCommand
    return await future
           ^^^^^^^^^^^^
chip.interaction_model.InteractionModelError: InteractionModelError: InvalidCommand (0x85)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/mobly/base_test.py", line 818, in exec_one_test
    test_method()
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 1671, in async_runner
    return _async_runner(body, self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/python_testing/scripts/sdk/matter_testing_support.py", line 1659, in _async_runner
    return asyncio.run(runner_with_timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
    return await fut
           ^^^^^^^^^
  File "/root/python_testing/scripts/sdk/TC_DEM_2_8.py", line 180, in test_TC_DEM_2_8
    await self.send_request_constraint_based_forecast(constraintList, cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization, expected_status=Status.ConstraintError)
  File "/root/python_testing/scripts/sdk/TC_DEMTestBase.py", line 170, in send_request_constraint_based_forecast
    asserts.assert_equal(e.status, expected_status, "Unexpected error returned")
  File "/usr/local/lib/python3.12/dist-packages/mobly/asserts.py", line 66, in assert_equal
    _call_unittest_assertion(
  File "/usr/local/lib/python3.12/dist-packages/mobly/asserts.py", line 50, in _call_unittest_assertion
    raise signals.TestFailure(my_msg, extras=extras)
mobly.signals.TestFailure: Details=<Status.InvalidCommand: 133> != <Status.ConstraintError: 135> Unexpected error returned, Extras=None`

### PR DESCRIPTION
**Pull request by PeterC1965**
Addresses various DEM testing issues:

- Fixes #34954
- Fixes #34960 
- Fixes #34962
- Fixes #34297

- https://github.com/CHIP-Specifications/chip-test-plans/issues/4415
- https://github.com/CHIP-Specifications/chip-test-plans/issues/4482

- Adds a new TC_DEM_2.10.py script for checking Q Quality and that MatterReporting callback is called when expected (found and fixed some issues here!)

This should be cross-reviewed against the DEM Test Plan change PR: 
https://github.com/CHIP-Specifications/chip-test-plans/pull/4543

### CODE REVIEW

File: src/python_testing/TC_DEM_2_10.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    def desc_TC_DEM_2_10(self) -> str:
+        """Return a description of this test."""
+        return "4.1.3. [TC-DEM-2.10] This test case verifies attributes of the Device Energy Mangement cluster server having the Q quality."
+
+    def pics_TC_DEM_2_10(self):
+        """Return the PICS definitions associated with this test."""
+        pics = [
+            # Depends on Feature 05 (ForecastAdjustment) & Feature 02 (StateForecastReporting)
+            "DEM.S.F05", "DEM.S.F02"
```
Comment (jamesharrow) [inline src/python_testing/TC_DEM_2_10.py:57]: Should be ok with PFR as well

File: src/python_testing/TC_DEM_2_10.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            TestStep("6b", "TH reads from the DUT the OptOutState",
+                     "Value has to be 0x02 (LocalOptOut)"),
+            TestStep("7", "TH sends command ModifyForecastRequest with ForecastID=Forecast.ForecastID, SlotAdjustments[0].{SlotIndex=0, Duration=Forecast.Slots[0].MinDurationAdjustment}, Cause=GridOptimization",
+                     "Verify DUT responds w/ status SUCCESS(0x00)"),
+            TestStep("8", "TH counts all report transactions with an attribute report for the Forecast attribute over the next Forecast.Slots[0].MinDurationAdjustment}",
+                     "TH verifies that numberOfReportsReceived <= 2 + Forecast.Slots[0].MinDurationAdjustment}"),
+            TestStep("9", "Cancel the subscription to the Forecast attribute",
+                     "The subscription is cancelled successfully"),
+            TestStep("10", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TESTEVENT_TRIGGERKEY and EventTrigger field set to PIXIT.DEM.TESTEVENTTRIGGER for Forecast Adjustment Test Event Clear",
+                     "Verify DUT responds w/ status SUCCESS(0x00)"),
```
Comment (jamesharrow) [inline src/python_testing/TC_DEM_2_10.py:92]: Need to add PowerAdjustment tests as well.

File: src/python_testing/TC_DEM_1_1.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
+#
```
Comment (jamesharrow) [inline src/python_testing/TC_DEM_1_1.py:1]: We have just removed this from the test plan :( so we don't need this now...

File: src/python_testing/TC_DEMTestBase.py [hunk 4]
```diff
@@ -24,6 +24,14 @@
 
 logger = logging.getLogger(__name__)
 
+s_feature_strs = {Clusters.DeviceEnergyManagement.Bitmaps.Feature.kPowerAdjustment:           "kPowerAdjustment",
```
Comment (jamesharrow) [inline src/python_testing/TC_DEMTestBase.py:27]: We can just use .name here

File: src/python_testing/TC_DEMTestBase.py [hunk 5]
```diff
@@ -38,6 +46,26 @@ async def check_dem_attribute(self, attribute, expected_value, endpoint: int = N
...
 
+    async def validate_feature_map(self, must_have_features, must_not_have_features):
+        feature_map = await self.read_dem_attribute_expect_success(attribute="FeatureMap")
+        for must_have_feature in must_have_features:
+            asserts.assert_true(feature_map & must_have_feature,
+                                f"{s_feature_strs[must_have_feature]} must be set but is not. feature_map 0x{feature_map:x}")
+
+        for must_not_have_feature in must_not_have_features:
+            asserts.assert_false(feature_map & must_not_have_feature,
+                                 f"{s_feature_strs[must_not_have_feature]} is not allowed to be set. feature_map 0x{feature_map:x}")
```
Comment (jamesharrow) [inline src/python_testing/TC_DEMTestBase.py:57]:
```suggestion
                                 f"{must_not_have_feature.name} is not allowed to be set. feature_map 0x{feature_map:x}")
```

File: src/python_testing/TC_DEMTestBase.py (hunk 4 above)
Comment (jamesharrow) [reply src/python_testing/TC_DEMTestBase.py:27]: Think this can be removed - see below

**Comment by jamesharrow** [comment]
I've reviewed Test cases 2.2 - 2.9 and ensured that they align with the test plan PR (https://github.com/CHIP-Specifications/chip-test-plans/pull/4543) fixing up minor issues. I think they all look good now - Thanks P!

I just need to look at 2.10 (Q Quality tests) to tweak this test case so it covers both Forecast and PowerAdjustment attributes.

File: src/python_testing/TC_DEM_2_10.py (hunk 1 above)
Comment (jamesharrow) [reply src/python_testing/TC_DEM_2_10.py]: Re-worked.

File: src/python_testing/TC_DEM_2_10.py (hunk 2 above)
Comment (jamesharrow) [reply src/python_testing/TC_DEM_2_10.py:105]: Reworked

**Review by doru91 — APPROVED** [review]

**Review by jamesharrow — APPROVED** [review]
Looks good to me now

**Review by hasty — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**